apollo-client/src/main/java/com/ctrip/framework/apollo/internals/K8sConfigMapConfigRepository.java

```diff
@@ -0,0 +1,308 @@
+import com.google.gson.Gson;
+import com.google.gson.reflect.TypeToken;
+import org.slf4j.Logger;
+
+import java.lang.reflect.Type;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Properties;
+
+/**
+ * @author dyx1234
+ */
+public class K8sConfigMapConfigRepository extends AbstractConfigRepository
+        implements RepositoryChangeListener {
+    private static final Logger logger = DeferredLoggerFactory.getLogger(K8sConfigMapConfigRepository.class);
+    private final String namespace;
+    private String configMapName;
+    private String configMapKey;
+    private final String configMapNamespace;
+    private final ConfigUtil configUtil;
+    private final KubernetesManager kubernetesManager;
+    private volatile Properties configMapProperties;
+    private volatile ConfigRepository upstream;
+    private volatile ConfigSourceType sourceType = ConfigSourceType.CONFIGMAP;
+
+    /**
+     * Constructor
+     *
+     * @param namespace the namespace
+     */
+    public K8sConfigMapConfigRepository(String namespace) {
+        this(namespace, (ConfigRepository) null);
+    }
+
+    public K8sConfigMapConfigRepository(String namespace, KubernetesManager kubernetesManager) {
+        this.namespace = namespace;
+        configUtil = ApolloInjector.getInstance(ConfigUtil.class);
+        kubernetesManager = ApolloInjector.getInstance(KubernetesManager.class);
+        configMapNamespace = configUtil.getConfigMapNamespace();
+        this.kubernetesManager = kubernetesManager;
+
+        this.setConfigMapKey(configUtil.getCluster(), namespace);
+        this.setConfigMapName(configUtil.getAppId(), false);
+        this.setUpstreamRepository(upstream);
+    }
+
+    public K8sConfigMapConfigRepository(String namespace, ConfigRepository upstream) {
+        this.namespace = namespace;
+        configUtil = ApolloInjector.getInstance(ConfigUtil.class);
+        kubernetesManager = ApolloInjector.getInstance(KubernetesManager.class);
+        configMapNamespace = configUtil.getConfigMapNamespace();
+
+        this.setConfigMapKey(configUtil.getCluster(), namespace);
+        this.setConfigMapName(configUtil.getAppId(), false);
+        this.setUpstreamRepository(upstream);
+    }
+
+    void setConfigMapKey(String cluster, String namespace) {
+        // TODO 兜底key怎么设计不会冲突(cluster初始化时已经设置了层级)
+        // cluster: 用户定义>idc>default,所以已经不需要额外层级设置了
+        if (StringUtils.isBlank(cluster)) {
+            configMapKey = Joiner.on(ConfigConsts.CLUSTER_NAMESPACE_SEPARATOR).join("default", namespace);
+            return;
+        }
+        configMapKey = Joiner.on(ConfigConsts.CLUSTER_NAMESPACE_SEPARATOR).join(cluster, namespace);
+    }
+
+    public String getConfigMapKey() {
+        return configMapKey;
+    }
+
+    public String getConfigMapName() {
+        return configMapName;
+    }
+
+    void setConfigMapName(String appId, boolean syncImmediately) {
+        configMapName = appId;
```

Comment:
It may be better to add a prefix to distinguish it from the configmap of the business, such as apollo-configcache-xxx